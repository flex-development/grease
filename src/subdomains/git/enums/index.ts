/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
