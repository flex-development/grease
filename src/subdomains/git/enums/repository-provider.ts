/**
 * @file Enums - RepositoryProvider
 * @module grease/git/enums/RepositoryProvider
 */

/**
 * Repository provider names.
 *
 * @enum {Lowercase<string>}
 */
enum RepositoryProvider {
  BITBUCKET = 'bitbucket',
  GITHUB = 'github',
  GITLAB = 'gitlab'
}

export default RepositoryProvider
