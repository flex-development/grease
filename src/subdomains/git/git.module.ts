/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { GitService } from './providers'

/**
 * Global git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  providers: [GitService, LoggerService, ValidationService]
})
class GitModule {}

export default GitModule
