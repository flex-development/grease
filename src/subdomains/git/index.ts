/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './options'
export * from './providers'
export type * from './types'
