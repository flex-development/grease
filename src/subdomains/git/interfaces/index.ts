/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
