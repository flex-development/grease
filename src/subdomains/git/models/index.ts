/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
