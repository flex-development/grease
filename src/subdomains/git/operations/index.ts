/**
 * @file Entry Point - Operations
 * @module grease/git/operations
 */

export { default as TagOperationHandler } from './tag.handler'
export { default as TagOperation, type TagOperationDTO } from './tag.operation'
