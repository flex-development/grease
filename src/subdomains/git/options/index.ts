/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitCommitOptions } from './git-commit.options'
export { default as GitTagOptions } from './git-tag.options'
export { default as GitOptions } from './git.options'
