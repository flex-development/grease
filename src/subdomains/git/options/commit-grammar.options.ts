/**
 * @file Options - CommitGrammarOptions
 * @module grease/git/options/CommitGrammarOptions
 */

import { RepositoryProvider } from '#src/git/enums'
import {
  get,
  isString,
  select,
  sort,
  unique,
  type Partial
} from '@flex-development/tutils'
import {
  IsArray,
  IsEnum,
  IsString,
  MinLength
} from 'class-validator'

/**
 * Commit parser grammar options.
 *
 * @class
 */
class CommitGrammarOptions {
  /**
   * Reference action keywords.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} actions
   */
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  public actions: string[]

  /**
   * Issue reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} issues
   */
  @IsArray()
  @IsString({ each: true })
  public issues: string[]

  /**
   * Pull request reference prefixes.
   *
   * @default
   *  based on provider
   *
   * @public
   * @instance
   * @member {string[]} pr
   */
  @IsArray()
  @IsString({ each: true })
  public pr: string[]

  /**
   * Repository provider name.
   *
   * @default RepositoryProvider.GITHUB
   *
   * @public
   * @instance
   * @member {RepositoryProvider} provider
   */
  @IsEnum(RepositoryProvider)
  public provider: RepositoryProvider

  /**
   * Tag prefix.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} tagprefix
   */
  @IsString()
  public tagprefix: string

  /**
   * Create a new options object.
   *
   * @param {Partial<CommitGrammarOptions>?} [opts] - Commit grammar options
   */
  constructor(opts?: Partial<CommitGrammarOptions>) {
    this.actions = select(get(opts, 'actions', []), isString)
    this.issues = select(get(opts, 'issues', []), isString)
    this.pr = select(get(opts, 'pr', []), isString)
    this.provider = get(opts, 'provider', RepositoryProvider.GITHUB)
    this.tagprefix = get(opts, 'tagprefix', '')

    // add common provider actions and prefixes
    this.actions.push('close', 'closed', 'closes', 'fix', 'fixed')
    this.issues.push('#')
    this.provider !== RepositoryProvider.GITLAB && this.pr.push('#')

    // add actions and prefixes based on provider
    switch (this.provider) {
      case RepositoryProvider.BITBUCKET:
        this.actions.push(
          'closing',
          'fixes',
          'fixing',
          'resolve',
          'resolved',
          'resolves',
          'resolving'
        )
        this.issues.unshift('JRA-', 'JIRA-')
        break
      case RepositoryProvider.GITLAB:
        this.actions.push('closing', 'fixes', 'fixing')
        this.pr.push('!')
        break
      default:
        this.actions.push('fixes', 'resolve', 'resolved', 'resolves')
        this.issues.push('GH-')
        break
    }

    // ensure actions and prefixes are unique
    this.actions = sort(unique(this.actions), (a, b) => b.length - a.length)
    this.issues = unique(this.issues)
    this.pr = unique(this.pr)
  }
}

export default CommitGrammarOptions
