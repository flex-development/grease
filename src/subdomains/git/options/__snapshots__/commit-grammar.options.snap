// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.BITBUCKET > should set #actions 1`] = `
[
  "resolving",
  "resolved",
  "resolves",
  "closing",
  "resolve",
  "closed",
  "closes",
  "fixing",
  "close",
  "fixed",
  "fixes",
  "fix",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.BITBUCKET > should set #issues 1`] = `
[
  "JRA-",
  "JIRA-",
  "#",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.BITBUCKET > should set #pr 1`] = `
[
  "#",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITHUB > should set #actions 1`] = `
[
  "resolved",
  "resolves",
  "resolve",
  "closed",
  "closes",
  "close",
  "fixed",
  "fixes",
  "fix",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITHUB > should set #issues 1`] = `
[
  "#",
  "GH-",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITHUB > should set #pr 1`] = `
[
  "#",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITLAB > should set #actions 1`] = `
[
  "closing",
  "closed",
  "closes",
  "fixing",
  "close",
  "fixed",
  "fixes",
  "fix",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITLAB > should set #issues 1`] = `
[
  "#",
]
`;

exports[`unit:git/options/CommitGrammarOptions > constructor > RepositoryProvider.GITLAB > should set #pr 1`] = `
[
  "!",
]
`;
